### ISSUE TITLE
Proposal: rule to check unwanted nested function calls

### ISSUE THREAD

**Issue by kunagpal** [eslint/eslint#8707]
**Please describe what the rule should do:**
Check if a function is being called within another function, with exactly the same arguments as the parent function.

**What category of rule is this? (place an "X" next to just one item)**

- [x] Enforces code style
- [ ] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// 1. Here, cb is a callback in the current scope
function done (err, results) {
    if (err) {
        return cb(err);
    }

    cb(null, results); // done can be replaced with cb directly.
}

// 2. Here, the inner function is being called with exactly the same arguments as its enclosing function. No other statements exist, just the innerFunction call.
function (foo, bar) {
    innerFunction(foo, bar);
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
1. This rule introduces and aims to enforce generic code style that is agnostic of the JavaScript environment.
2. If introduced, this rule will help cut down on unwanted function calls and facilitate writing compact code.

**Comment by ljharb** [comment]
This can often be a valid technique to abstract around the inner function - perhaps its `.length` isn't the same as the wrapper, or its `.name`, or its `.toString` - or perhaps you just want to insulate yourself from changes to its API (by giving you a single place to refactor if it changes).

**Comment by Pyrolistical** [comment]
You could run into a bug with the identity function, if it isn't really an identity function.

Let's say you have calling code: `foo(1, 2, 3)`, where `foo` is 
```
function foo(a, b) {
  bar(a, b);
}
```

Replacing the call to foo with bar can be a bug if bar has a 3rd optional parameter. I know this is bad code and maybe a rule like this is worth introducing bugs so they get fixed.

**Comment by platinumazure** [comment]
Added :-1: for the reasons stated above.

Note that if this isn't accepted for ESLint core, you are welcome to make a plugin for this (or find another plugin to which this proposed rule could be added).

**Comment by not-an-aardvark** [comment]
It looks like the team isn't really in favor of this issue, so I'm going to go ahead and close it. Thanks for the proposal regardless!
